### ISSUE TITLE
fix risearch queries for full reindexing

### ISSUE THREAD

**Issue by jmartin-sul** [sul-dlss/argo#380]
some background:
- As I noticed earlier, and as @rosy1280 pointed out to me today, the argo-prod-b index is currently about 200k records short of the argo-prod index.  Rosy says these are mostly Google books, but she can clarify the characteristics of the shortfall whenever she gets a chance (she had other observations I've forgotten).
- IIRC from when we had to fully reindex Argo last fall, the druids returned by Fedora's risearch interface did not provide a complete list for what needed to be indexed.  I have vague memories of Willy saying that he needed to query Fedora's underlying MySQL database to get the full list of PIDs for reindexing, but there was no explanation as to why.
- Fedora's risearch functionality uses Fedora's triple store index, which we turned off some time ago (prior to last fall's full reindex).  I believe it was for performance reasons, but @LynnMcRae can clarify there.  So risearch will return stale results on our Fedora instances.
- AFAICT, Fedora's underlying relational database does not contain any usable info on object type.  The available tables in the Fedora DB are:

```
mysql> show tables;
+---------------------+
| Tables_in_fedora    |
+---------------------+
| datastreamPaths     |
| dcDates             |
| doFields            |
| doRegistry          |
| fcrepoRebuildStatus |
| modelDeploymentMap  |
| objectPaths         |
| pidGen              |
+---------------------+
```

```
mysql> describe doRegistry;
+---------------+--------------+------+-----+---------+-------+
| Field         | Type         | Null | Key | Default | Extra |
+---------------+--------------+------+-----+---------+-------+
| doPID         | varchar(64)  | NO   | PRI | NULL    |       |
| systemVersion | int(11)      | NO   |     | 0       |       |
| ownerId       | varchar(64)  | YES  |     | NULL    |       |
| objectState   | varchar(1)   | NO   |     | A       |       |
| label         | varchar(255) | YES  |     |         |       |
+---------------+--------------+------+-----+---------+-------+
```

```
mysql> describe doFields;
+---------------+--------------+------+-----+---------+-------+
| Field         | Type         | Null | Key | Default | Extra |
+---------------+--------------+------+-----+---------+-------+
| pid           | varchar(64)  | NO   | MUL | NULL    |       |
| label         | varchar(255) | YES  |     | NULL    |       |
| state         | varchar(1)   | NO   |     | A       |       |
| ownerId       | varchar(64)  | YES  |     | NULL    |       |
| cDate         | bigint(20)   | NO   |     | NULL    |       |
| mDate         | bigint(20)   | NO   |     | NULL    |       |
| dcmDate       | bigint(20)   | YES  |     | NULL    |       |
| dcTitle       | text         | YES  |     | NULL    |       |
| dcCreator     | text         | YES  |     | NULL    |       |
| dcSubject     | text         | YES  |     | NULL    |       |
| dcDescription | text         | YES  |     | NULL    |       |
| dcPublisher   | text         | YES  |     | NULL    |       |
| dcContributor | text         | YES  |     | NULL    |       |
| dcDate        | text         | YES  |     | NULL    |       |
| dcType        | text         | YES  |     | NULL    |       |
| dcFormat      | text         | YES  |     | NULL    |       |
| dcIdentifier  | text         | YES  |     | NULL    |       |
| dcSource      | text         | YES  |     | NULL    |       |
| dcLanguage    | text         | YES  |     | NULL    |       |
| dcRelation    | text         | YES  |     | NULL    |       |
| dcCoverage    | text         | YES  |     | NULL    |       |
| dcRights      | text         | YES  |     | NULL    |       |
+---------------+--------------+------+-----+---------+-------+
```

```
mysql> select count(dcType) from doFields;
+---------------+
| count(dcType) |
+---------------+
|             0 |
+---------------+

mysql> select distinct dcType from doFields;
+--------+
| dcType |
+--------+
| NULL   |
+--------+

mysql> select count(pid) from doFields;
+------------+
| count(pid) |
+------------+
|     891662 |
+------------+
```

Fortunately, we do have usable type info in the old index (current Argo prod's index).  However, you can imagine how it'd be problematic if you needed to rebuild your current index without the luxury of an existing index.  Then you'd have to index everything once, find the objects of the specific types you want to index first, then re-run the indexing with the proper ordering.  If Fedora's triple store index were enabled, you could actually query your data store and get rich results based on things like object type.  I have no idea whether performance problems make that infeasible, but it seemed worth griping about anyway.

Another thing that's obviously problematic is that it makes it that much harder to do indexing from Ruby code running on the Argo VM.  You'd either need to have that code talk to Fedora's underlying RDBMS instance, or you'd have to export the PID list by hand and bring the file over.  The former sounds ugly architecturally, and the latter sounds like a pain, though not a great one.  The part about not being able to query Fedora by object type in our current environment seems much worse.

TL;DR:
- argo-prod-b index shortfall appears to be due to stale query results (at first pass, need to diff the PID list from MySQL against the ones returned by the risearch queries, vs what's indexed for argo-prod-b).
- rebuilding an index from scratch with our current fedora setup and no existing index would, AFAICT, require two full reindexing passes.  which is around a work week, since reindexing currently takes 2.5 to 3 days on prod.  totally possible i'm missing something there.  i guess you could also crawl fedora's file system and load the foxml and get type info that way?

**Comment by rosy1280** [comment]
I can't be certain that they are google books, that was my best guess when i initially started looking.  But without tags in the facets I can't be certain.  I did note that there were no google books workflows appearing in Argo -- but perhaps that's because not all of the workflow objects were indexed?

**Comment by jmartin-sul** [comment]
i re-ran some queries, to get the count for what dor-prod's risearch returns, vs what's in argo-prod-b's solr index, vs the list from sql.  since the lists are so different in length, i don't know how productive it'd be to diff and see what individual druids are missing, but i'm happy to do that if it seems useful.  anyway:

```
2.2.0 :048 >   pid_lists_for_full_reindex.map(&:length).reduce(:+)
 => 672663 

2.2.0 :055 >   solr_pids.length
 => 670394 

2.2.0 :057 >   pid_gatherer.all_pids.length
I, [2015-12-16T11:35:46.581939 #15333]  INFO -- : found 672619 pids in fedora (all pids, unfiltered)
 => 672619 

2.2.0 :059 >   pid_lists_for_full_reindex.reduce(:+).uniq.length
 => 672662 
```

here's an interesting thing:  "`all_pids`" comes up with 43 fewer objects than are in the full reindexing list (squashed into one list with dupes removed, 672619 vs 672662).  `all_pids` iterates over all the pids in fedora, whereas pid_lists_for_full_reindex gets lists by object type, with a couple of hardcoded objects like uber APO and hydrus uber APO.  so that discrepancy alone makes me distrust risearch.

for reference, the mysql list has 891662 entries.

also:  when we indexed last fall, we had about 500k objects in fedora.  but if risearch was turned off before that, you'd think we'd get fewer than 500k objects in the current index, but we get almost 200k more than that.  also, @drh-stanford discovered that risearch doesn't seem to be turned off, according to the settings he found, and @LynnMcRae can't find a jira issue indicating it was ever disabled.  so either we're querying fedora incorrectly, or risearch doesn't give complete results.  @drh-stanford suggests asking the fedora tech mailing list and seeing what help they can provide.

in the meantime, i'd propose just diffing the current argo-prod-b solr pid list against the mysql pid list, and using that to represent "all pids" (really all remaining unindexed pids).  then query current prod solr for the things that need to be indexed first (workflows, APOs, etc).  use those hand assembled lists to create a new list of lists for indexing, without the stuff that's already in solr.  start an indexing run based on that by feeding it to the batching mechanism.

**Comment by drh-stanford** [comment]
i'm seeing 891572 objects in the existing solr production index. the mysql index and our existing solr index do not match in object count: 891662 (fedora) > 891572 (solr). so maybe we've deleted/purged ~100 objects and fedora keeps a record of delete objects, or our solr index is out of date. otherwise than expecting a ~100 objects not to resolve, i think your proposal is good.

**Comment by LynnMcRae** [comment]
I was expecting to find a small number of orphaned solr docs in prod if anyone purged a DOR object directly in the Fedora admin interface and not through our service that cleans up the workspace, workflows and solr.  Maybe that’s not really so since we had to rebuild the index just a year ago and have been disciplined about decommission and purge since.  100 fewer in solr is more of a mystery, perhaps damaged objects that get errors in indexing? Objects we’d not likely care about but should identify.

Lynn

**Comment by jmartin-sul** [comment]
queries used to get PIDs from the current prod solr index by object type:
```
q=objectType_facet:"workflow"
q=objectType_facet:"agreement"
q=objectType_facet:"adminPolicy"
q=objectType_facet:"collection"
q=objectType_facet:"set"
```

results obtained using this method in the prod rails console (sort of a knock off of https://github.com/sul-dlss/argo/pull/246/files#diff-673f632e8c23d7491d779e7f123ca426R41):
```
def solr_pids(q = '*:*')
  puts "querying solr for indexed pids, (q=#{q})..."

  start = 0
  solr_pids = []
  resp = Dor::SearchService.query(q, :sort => 'id asc', :rows => 1000, :start => start, :fl => ['id'])
  while resp.docs.length > 0
    solr_pids += resp.docs.map { |doc| doc['id'] }
    start += 1000
    resp = Dor::SearchService.query(q, :sort => 'id asc', :rows => 1000, :start => start, :fl => ['id'])
  end
  puts "found #{solr_pids.length} pids in solr (unfiltered)"
  solr_pids
end
```

dumped the results as JSON to text files, will transfer those to argo-prod-b and load them up there.

**Comment by jmartin-sul** [comment]
some quick notes before we leave for winter break:
* i tried to index the things that were left out of the initial round of argo-prod-b indexing.  in pry, i 
 * created a `PidGatherer` instance where i loaded up:
  * `all_pids` with the full dump from prod mysql
  * workflow, agreement, apo, collection, and set lists with lists from the current prod solr index
  * `solr_pids` with the dump i'd done of argo-prod-b earlier in the day
  * all the other setable lists with [] (the hydrus specific ones)
 * took that `PidGatherer` instance and used it to calculate `pid_lists_for_unindexed`, then fed that to a `BulkIndexer` instance.

that got somewhere around 200k pids.  those finished indexing in about 10 hrs (around 6:30 pm to around 8:30 am).  i believe argo-prod-b is still about 2000 items short of the prod index.

i have all the generated pid lists, and logs of the two important pry sessions (the export from current prod and the indexer list build on prod-b).  additions and corrections after winter break (but i have to head to the airport in about 3 hours, so this is as far as i can get for now).

**Comment by drh-stanford** [comment]
is this moot now?

**Comment by jmartin-sul** [comment]
i think this is still a problem, though not an urgent one.  `risearch` doesn't work as advertised (possibly due to the way we have it configured), and so pid gathering for a full reindex is still a chore that involves a decent amount of manual legwork (see https://github.com/sul-dlss/argo/pull/483 and https://github.com/sul-dlss/argo/pull/482, which are intended to make it a bit easier given the current fedora query problem).  not an untenable situation since we're not doing tons of full reindexing on an regular basis, but worth leaving in the backlog, IMO.

we could close this and open a new ticket specifically about either fixing `risearch` or making the building of a full reindex list more automated (i thought there might be a ticket like that, but looks like this is it).  i'd be in favor of fixing `risearch` if possible, since it'd allow much more query flexibility, and wouldn't require a pre-existing solr index, which might not be a thing we have available (if, e.g., we're reindexing because the index got wiped).

**Comment by drh-stanford** [comment]
ok, i renamed the issue and removed the +1

**Comment by LynnMcRae** [comment]
We may have a situation that ETD relied on risearch to integrate objects when they were submitted. I have not yet confirmed this. There is an ETD report report that uses risearch, so it's a mystery why that still seems to be working yet I cannot reproduce its results. Still working on this.

**Comment by jcoyne** [comment]
Going to close this by saying that our current path is to not rely on risearch in order to do a full-reindex.
